File: src/template-parser.ts
import type { Delimiters, RenderOptions } from "./types.js";
import {
  escapeRegExp,
  expressionClose,
  expressionOpen,
  makeDelimiters,
  readResource,
  variableClose,
  variableOpen
} from "./util.js";
import { Scanner, readTagInfoFromString, tagInnerContent } from "./scanner.js";
import { validate } from "./validator.js";

interface BlockInfo {
  super: boolean;
  content: string;
}

type Blocks = Record<string, BlockInfo | undefined>;
type Defaults = Record<string, string> | undefined;

class Buffer {
  value = "";
  append(s: string | undefined): void {
    if (s != null) this.value += s;
  }
  toString(): string {
    return this.value;
  }
}

function tagInfo(tagStr: string, delimiters: Delimiters) {
  return readTagInfoFromString(tagStr, delimiters);
}

export function getTagParams(tagId: string, blockStr: string, opts: RenderOptions = {}): string {
  const d = makeDelimiters(opts);
  const info = tagInfo(blockStr, d);
  if (info.tagType !== "expr" || info.tagName !== tagId) return "";
  return info.args.join(" ").trim();
}

export function parseDefaults(defaultParts?: string[]): Defaults {
  if (!defaultParts || defaultParts.length === 0) return undefined;
  const joined = defaultParts.join(" ");
  const defaults: Record<string, string> = {};
  const re = /([^=\s]+)\s*=\s*"([^"]*)"/g;
  let match: RegExpExecArray | null;
  while ((match = re.exec(joined))) defaults[match[1]] = match[2];
  return defaults;
}

export function splitIncludeTag(tagStr: string, opts: RenderOptions = {}): string[] {
  const params = getTagParams("include", tagStr.replace(/\\/g, "/"), opts);
  return params.match(/"[^"]*"|[^\s]+/g) ?? [];
}

export function getParent(tagStr: string, opts: RenderOptions = {}): string {
  const template = getTagParams("extends", tagStr, opts);
  return template.startsWith('"') && template.endsWith('"') ? template.slice(1, -1) : template;
}

function classifyTag(tagStr: string, delimiters: Delimiters) {
  const info = tagInfo(tagStr, delimiters);
  return {
    info,
    include: info.tagType === "expr" && info.tagName === "include",
    extends: info.tagType === "expr" && info.tagName === "extends",
    block: info.tagType === "expr" && info.tagName === "block",
    endblock: info.tagType === "expr" && info.tagName === "endblock",
    blockName: info.tagType === "expr" && info.tagName === "block" ? info.args.join(" ").trim() : undefined,
    superTag: info.tagType === "filter" && info.tagValue.trim() === "block.super"
  };
}

function writeTag(superTag: boolean, existingBlock: BlockInfo | undefined, blocksToClose: number, omitCloseTag: boolean | undefined): boolean {
  return superTag || (!existingBlock && blocksToClose > (omitCloseTag ? 1 : 0));
}

function processIncludes(tagStr: string, blocks: Blocks, opts: RenderOptions): string {
  const params = splitIncludeTag(tagStr, opts);
  const source = (params[0] ?? "").replace(/"/g, "");
  const defaults = parseDefaults(params.slice(2));
  return preprocessTemplateFile(source, opts, blocks, defaults);
}

export function consumeBlock(scanner: Scanner, opts: RenderOptions, buf?: Buffer, blocks: Blocks = {}, omitCloseTag = false): boolean {
  let blocksToClose = 1;
  let hasSuper = false;
  const d = scanner.delimiters;

  while (blocksToClose > 0 && !scanner.eof()) {
    if (startsTag(scanner)) {
      const tagStr = scanner.readTagContent();
      const tag = classifyTag(tagStr, d);
      const existingBlock = tag.blockName ? blocks[tag.blockName] : undefined;

      if (buf) {
        if (tag.include) buf.append(processIncludes(tagStr, blocks, opts));
        else if (writeTag(tag.superTag, existingBlock, blocksToClose, omitCloseTag)) buf.append(tagStr);
      }

      if (existingBlock && tag.blockName) {
        consumeBlock(scanner, opts);
        const nestedBlocks = { ...blocks };
        delete nestedBlocks[tag.blockName];
        consumeBlock(new Scanner(existingBlock.content, d), opts, buf, nestedBlocks);
      } else if (tag.block) {
        blocksToClose += 1;
      } else if (tag.endblock) {
        blocksToClose -= 1;
      }
      hasSuper = hasSuper || tag.superTag;
    } else {
      const ch = scanner.readChar();
      buf?.append(ch);
    }
  }
  return hasSuper;
}

export function rewriteSuper(block: string, parentContent: string, opts: RenderOptions = {}): string {
  const d = makeDelimiters(opts);
  const superTag = `${escapeRegExp(variableOpen(d))}\\s*block\\.super\\s*${escapeRegExp(variableClose(d))}`;
  return block.replace(new RegExp(superTag, "g"), parentContent);
}

export function readBlock(scanner: Scanner, blockTag: string, blocks: Blocks, opts: RenderOptions): Blocks {
  const d = scanner.delimiters;
  const blockName = classifyTag(blockTag, d).blockName ?? "";
  const existingBlock = blocks[blockName];
  if (existingBlock?.super) {
    const childContent = existingBlock.content;
    const parentContent = new Buffer();
    const hasSuper = consumeBlock(scanner, opts, parentContent, blocks, true);
    return { ...blocks, [blockName]: { super: hasSuper, content: rewriteSuper(childContent, parentContent.toString(), opts) } };
  }
  if (existingBlock) {
    consumeBlock(scanner, opts);
    return blocks;
  }
  const buf = new Buffer();
  buf.append(blockTag);
  const hasSuper = consumeBlock(scanner, opts, buf, blocks);
  return { ...blocks, [blockName]: { super: hasSuper, content: buf.toString() } };
}

export function processBlock(scanner: Scanner, buf: Buffer, blockTag: string, blocks: Blocks, opts: RenderOptions): void {
  const d = scanner.delimiters;
  const blockName = classifyTag(blockTag, d).blockName ?? "";
  const childContent = blocks[blockName]?.content;
  if (childContent != null) {
    const parent = new Buffer();
    consumeBlock(scanner, opts, parent, blocks, true);
    buf.append(rewriteSuper(childContent, parent.toString(), opts));
  } else {
    buf.append(blockTag);
    consumeBlock(scanner, opts, buf, blocks);
  }
}

export function wrapInExpressionTag(s: string, opts: RenderOptions = {}): string {
  const d = makeDelimiters(opts);
  return `${expressionOpen(d)}${s}${expressionClose(d)}`;
}

export function wrapInVariableTag(s: string, opts: RenderOptions = {}): string {
  const d = makeDelimiters(opts);
  return `${variableOpen(d)}${s}${variableClose(d)}`;
}

export function trimVariableTag(s: string, opts: RenderOptions = {}): string {
  return tagInnerContent(s, makeDelimiters(opts));
}

export function trimExpressionTag(s: string, opts: RenderOptions = {}): string {
  return tagInnerContent(s, makeDelimiters(opts));
}

function unparseDefaults(defaults: Defaults): string | undefined {
  if (!defaults) return undefined;
  return Object.entries(defaults).map(([k, v]) => `${k}="${v}"`).join(" ").trim();
}

export function toExpressionString(tagName: string, args: string[], defaults: Defaults, opts: RenderOptions = {}): string {
  const defaultsString = tagName === "include" ? unparseDefaults(defaults) : undefined;
  const joined = `${tagName}${args.length > 0 ? ` ${args.join(" ")}` : ""}${defaultsString ? ` with ${defaultsString}` : ""}`;
  return wrapInExpressionTag(joined, opts);
}

export function addDefault(identifier: string, defaultValue: string): string {
  return `${identifier}|default:"${defaultValue}"`;
}

export function tryAddDefault(identifier: string, defaults: Defaults): string {
  return defaults?.[identifier] != null ? addDefault(identifier, defaults[identifier]) : identifier;
}

export function addDefaultsToVariableTag(tagStr: string, defaults: Defaults, opts: RenderOptions = {}): string {
  return wrapInVariableTag(tryAddDefault(trimVariableTag(tagStr, opts), defaults), opts);
}

export function addDefaultsToExpressionTag(tagStr: string, defaults: Defaults, opts: RenderOptions = {}): string {
  const d = makeDelimiters(opts);
  const info = tagInfo(tagStr, d);
  if (info.tagType !== "expr") return tagStr;
  const args = info.args.map((arg) => tryAddDefault(arg, defaults));
  return toExpressionString(info.tagName, args, defaults, opts);
}

export function preprocessTemplateString(template: string, opts: RenderOptions = {}, blocks?: Blocks, defaults?: Defaults): string {
  return readTemplate(template, opts, blocks, defaults, true);
}

export function preprocessTemplateFile(template: string | URL, opts: RenderOptions = {}, blocks?: Blocks, defaults?: Defaults): string {
  return readTemplate(template, opts, blocks, defaults, false);
}

export function preprocessTemplate(template: string | URL, opts: RenderOptions = {}): string {
  return typeof template === "string" && template.includes("\n") ? preprocessTemplateString(template, opts) : preprocessTemplateFile(template, opts);
}

function readTemplate(template: string | URL, opts: RenderOptions, blocks: Blocks = {}, defaults?: Defaults, stringMode = false): string {
  const d = makeDelimiters(opts);
  const source = stringMode ? String(template) : readTemplateFile(template, opts);
  const scanner = new Scanner(source, d);
  const buf = new Buffer();
  let parent: string | undefined;
  let currentBlocks = blocks;

  while (!scanner.eof()) {
    if (startsTag(scanner)) {
      const tagStr = scanner.readTagContent();
      const tag = classifyTag(tagStr, d);
      if (defaults && tag.info.tagType === "filter") {
        buf.append(addDefaultsToVariableTag(tagStr, defaults, opts));
      } else if (defaults && tag.info.tagType === "expr" && !tag.include) {
        buf.append(addDefaultsToExpressionTag(tagStr, defaults, opts));
      } else if (defaults && tag.include) {
        buf.append(processIncludes(addDefaultsToExpressionTag(tagStr, defaults, opts), currentBlocks, opts));
      } else if (tag.include) {
        buf.append(processIncludes(tagStr, currentBlocks, opts));
      } else if (tag.extends) {
        parent = getParent(tagStr, opts);
      } else if (parent && tag.block) {
        currentBlocks = readBlock(scanner, tagStr, currentBlocks, opts);
      } else if (tag.block) {
        processBlock(scanner, buf, tagStr, currentBlocks, opts);
      } else if (!parent) {
        buf.append(tagStr);
      }
    } else {
      const ch = scanner.readChar();
      if (!parent) buf.append(ch);
    }
  }

  return parent ? readTemplate(parent, opts, currentBlocks, defaults, false) : buf.toString();
}

function readTemplateFile(template: string | URL, opts: RenderOptions): string {
  validate(template, opts);
  return readResource(template, opts).content;
}

function startsTag(scanner: Scanner): boolean {
  const d = scanner.delimiters;
  return scanner.startsWith(d.tagOpen + d.filterOpen) || scanner.startsWith(d.tagOpen + d.tagSecond);
}

File: src/selmer/template_parser.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

from .scanner import Scanner, read_tag_info_from_string, tag_inner_content
from .types import Delimiters
from .util import (
    expression_close,
    expression_open,
    make_delimiters,
    read_resource,
    variable_close,
    variable_open,
)
from .validator import validate


@dataclass
class BlockInfo:
    super: bool
    content: str


Blocks = dict[str, BlockInfo]
Defaults = dict[str, str] | None


class Buffer:
    def __init__(self) -> None:
        self.value = ""

    def append(self, value: str | None) -> None:
        if value is not None:
            self.value += value

    def __str__(self) -> str:
        return self.value


def _tag_info(tag_str: str, delimiters: Delimiters):
    return read_tag_info_from_string(tag_str, delimiters)


def get_tag_params(tag_id: str, block_str: str, **opts: Any) -> str:
    d = make_delimiters(**opts)
    info = _tag_info(block_str, d)
    if getattr(info, "tag_type", None) != "expr" or getattr(info, "tag_name", None) != tag_id:
        return ""
    return " ".join(getattr(info, "args", [])).strip()


def parse_defaults(default_parts: list[str] | None) -> Defaults:
    if not default_parts:
        return None
    joined = " ".join(default_parts)
    return {match.group(1): match.group(2) for match in re.finditer(r'([^=\s]+)\s*=\s*"([^"]*)"', joined)}


def split_include_tag(tag_str: str, **opts: Any) -> list[str]:
    params = get_tag_params("include", tag_str.replace("\\", "/"), **opts)
    return re.findall(r'"[^"]*"|[^\s]+', params)


def get_parent(tag_str: str, **opts: Any) -> str:
    template = get_tag_params("extends", tag_str, **opts)
    return template[1:-1] if template.startswith('"') and template.endswith('"') else template


def _classify_tag(tag_str: str, delimiters: Delimiters) -> dict[str, Any]:
    info = _tag_info(tag_str, delimiters)
    is_expr = getattr(info, "tag_type", None) == "expr"
    is_filter = getattr(info, "tag_type", None) == "filter"
    tag_name = getattr(info, "tag_name", None)
    return {
        "info": info,
        "include": is_expr and tag_name == "include",
        "extends": is_expr and tag_name == "extends",
        "block": is_expr and tag_name == "block",
        "endblock": is_expr and tag_name == "endblock",
        "block_name": " ".join(getattr(info, "args", [])).strip() if is_expr and tag_name == "block" else None,
        "super_tag": is_filter and getattr(info, "tag_value", "").strip() == "block.super",
    }


def _write_tag(super_tag: bool, existing_block: BlockInfo | None, blocks_to_close: int, omit_close_tag: bool) -> bool:
    return super_tag or (existing_block is None and blocks_to_close > (1 if omit_close_tag else 0))


def _process_includes(tag_str: str, blocks: Blocks, opts: dict[str, Any]) -> str:
    params = split_include_tag(tag_str, **opts)
    source = (params[0] if params else "").replace('"', "")
    defaults = parse_defaults(params[2:])
    return preprocess_template_file(source, blocks=blocks, defaults=defaults, **opts)


def _starts_tag(scanner: Scanner) -> bool:
    d = scanner.delimiters
    return scanner.startswith(d.tag_open + d.filter_open) or scanner.startswith(d.tag_open + d.tag_second)


def consume_block(scanner: Scanner, buf: Buffer | None = None, blocks: Blocks | None = None, omit_close_tag: bool = False, **opts: Any) -> bool:
    blocks = blocks or {}
    blocks_to_close = 1
    has_super = False
    d = scanner.delimiters
    while blocks_to_close > 0 and not scanner.eof():
        if _starts_tag(scanner):
            tag_str = scanner.read_tag_content()
            tag = _classify_tag(tag_str, d)
            block_name = tag["block_name"]
            existing_block = blocks.get(block_name) if block_name else None
            if buf is not None:
                if tag["include"]:
                    buf.append(_process_includes(tag_str, blocks, opts))
                elif _write_tag(tag["super_tag"], existing_block, blocks_to_close, omit_close_tag):
                    buf.append(tag_str)
            if existing_block and block_name:
                consume_block(scanner, blocks=blocks, **opts)
                nested_blocks = dict(blocks)
                nested_blocks.pop(block_name, None)
                consume_block(Scanner(existing_block.content, d), buf, nested_blocks, **opts)
            elif tag["block"]:
                blocks_to_close += 1
            elif tag["endblock"]:
                blocks_to_close -= 1
            has_super = has_super or tag["super_tag"]
        else:
            ch = scanner.read_char()
            if buf is not None:
                buf.append(ch)
    return has_super


def rewrite_super(block: str, parent_content: str, **opts: Any) -> str:
    d = make_delimiters(**opts)
    super_tag = f"{re.escape(variable_open(d))}\\s*block\\.super\\s*{re.escape(variable_close(d))}"
    return re.sub(super_tag, parent_content, block)


def read_block(scanner: Scanner, block_tag: str, blocks: Blocks, **opts: Any) -> Blocks:
    block_name = _classify_tag(block_tag, scanner.delimiters)["block_name"] or ""
    existing = blocks.get(block_name)
    if existing and existing.super:
        child_content = existing.content
        parent = Buffer()
        has_super = consume_block(scanner, parent, blocks, True, **opts)
        updated = dict(blocks)
        updated[block_name] = BlockInfo(has_super, rewrite_super(child_content, str(parent), **opts))
        return updated
    if existing:
        consume_block(scanner, **opts)
        return blocks
    buf = Buffer()
    buf.append(block_tag)
    has_super = consume_block(scanner, buf, blocks, **opts)
    updated = dict(blocks)
    updated[block_name] = BlockInfo(has_super, str(buf))
    return updated


def process_block(scanner: Scanner, buf: Buffer, block_tag: str, blocks: Blocks, **opts: Any) -> None:
    block_name = _classify_tag(block_tag, scanner.delimiters)["block_name"] or ""
    child = blocks.get(block_name)
    if child is not None:
        parent = Buffer()
        consume_block(scanner, parent, blocks, True, **opts)
        buf.append(rewrite_super(child.content, str(parent), **opts))
    else:
        buf.append(block_tag)
        consume_block(scanner, buf, blocks, **opts)


def wrap_in_expression_tag(s: str, **opts: Any) -> str:
    d = make_delimiters(**opts)
    return f"{expression_open(d)}{s}{expression_close(d)}"


def wrap_in_variable_tag(s: str, **opts: Any) -> str:
    d = make_delimiters(**opts)
    return f"{variable_open(d)}{s}{variable_close(d)}"


def trim_variable_tag(s: str, **opts: Any) -> str:
    return tag_inner_content(s, make_delimiters(**opts))


def trim_expression_tag(s: str, **opts: Any) -> str:
    return tag_inner_content(s, make_delimiters(**opts))


def _unparse_defaults(defaults: Defaults) -> str | None:
    if not defaults:
        return None
    return " ".join(f'{k}="{v}"' for k, v in defaults.items()).strip()


def to_expression_string(tag_name: str, args: list[str], defaults: Defaults, **opts: Any) -> str:
    defaults_s = _unparse_defaults(defaults) if tag_name == "include" else None
    joined = f"{tag_name}{' ' + ' '.join(args) if args else ''}{' with ' + defaults_s if defaults_s else ''}"
    return wrap_in_expression_tag(joined, **opts)


def add_default(identifier: str, default: str) -> str:
    return f'{identifier}|default:"{default}"'


def try_add_default(identifier: str, defaults: Defaults) -> str:
    return add_default(identifier, defaults[identifier]) if defaults and identifier in defaults else identifier


def add_defaults_to_variable_tag(tag_str: str, defaults: Defaults, **opts: Any) -> str:
    return wrap_in_variable_tag(try_add_default(trim_variable_tag(tag_str, **opts), defaults), **opts)


def add_defaults_to_expression_tag(tag_str: str, defaults: Defaults, **opts: Any) -> str:
    info = _tag_info(tag_str, make_delimiters(**opts))
    if getattr(info, "tag_type", None) != "expr":
        return tag_str
    args = [try_add_default(arg, defaults) for arg in getattr(info, "args", [])]
    return to_expression_string(getattr(info, "tag_name"), args, defaults, **opts)


def preprocess_template_string(template: str, *, blocks: Blocks | None = None, defaults: Defaults = None, **opts: Any) -> str:
    return _read_template(template, True, blocks or {}, defaults, opts)


def preprocess_template_file(template: str, *, blocks: Blocks | None = None, defaults: Defaults = None, **opts: Any) -> str:
    return _read_template(template, False, blocks or {}, defaults, opts)


def preprocess_template(template: str, **opts: Any) -> str:
    return preprocess_template_string(template, **opts) if "\n" in str(template) else preprocess_template_file(template, **opts)


def _read_template(template: str, string_mode: bool, blocks: Blocks, defaults: Defaults, opts: dict[str, Any]) -> str:
    d = make_delimiters(**opts)
    source = str(template) if string_mode else _read_template_file(str(template), opts)
    scanner = Scanner(source, d)
    buf = Buffer()
    parent: str | None = None
    current_blocks = blocks
    while not scanner.eof():
        if _starts_tag(scanner):
            tag_str = scanner.read_tag_content()
            tag = _classify_tag(tag_str, d)
            if defaults and getattr(tag["info"], "tag_type", None) == "filter":
                buf.append(add_defaults_to_variable_tag(tag_str, defaults, **opts))
            elif defaults and getattr(tag["info"], "tag_type", None) == "expr" and not tag["include"]:
                buf.append(add_defaults_to_expression_tag(tag_str, defaults, **opts))
            elif defaults and tag["include"]:
                buf.append(_process_includes(add_defaults_to_expression_tag(tag_str, defaults, **opts), current_blocks, opts))
            elif tag["include"]:
                buf.append(_process_includes(tag_str, current_blocks, opts))
            elif tag["extends"]:
                parent = get_parent(tag_str, **opts)
            elif parent and tag["block"]:
                current_blocks = read_block(scanner, tag_str, current_blocks, **opts)
            elif tag["block"]:
                process_block(scanner, buf, tag_str, current_blocks, **opts)
            elif parent is None:
                buf.append(tag_str)
        else:
            ch = scanner.read_char()
            if parent is None:
                buf.append(ch)
    return _read_template(parent, False, current_blocks, defaults, opts) if parent else str(buf)


def _read_template_file(template: str, opts: dict[str, Any]) -> str:
    validate(template, **opts)
    content, _resource = read_resource(template, **opts)
    return content
